### PR TITLE
Fix of #989

### PR DESCRIPTION
**Pull request by cyberone**

### CODE REVIEW

**Comment by dmarkov** [comment]
Thanks, I'll find someone to review this PR soon

**Comment by dmarkov** [comment]
@longtimeago please review this PR

**Comment by longtimeago** [comment]
Several remarks:
1) We need to reproduce the issue with test first
2) It would be good to log some warning when skipping test due to the BindException.
3) Build is failed because multiple code quality errors
4) I guess, BindException can be thrown in any test which uses ```MkGrizzlyContainer```, right?

**Comment by cyberone** [comment]
@longtimeago 
1) the bug is reproducable, just occupy all ports (much overhead).
2) done
3) done, sorry, I thought I checked style before push
4) Yes, I left a puzzle.

**Comment by longtimeago** [comment]
@cyberone Thanks for 2,3,4
For the first point - the right way is to add a test which explicitly throws BindException and check that test is passed and warning is logged

**Comment by cyberone** [comment]
@longtimeago but it is testing a test, istn't it?

File: src/test/java/com/jcabi/github/RtForkTest.java [hunk 1]
```diff
@@ -108,4 +123,28 @@ private static JsonObject fork(final String organization) {
             .add("name", "nm")
             .build();
     }
+
+    private static class SkipBindException implements MethodRule {
```
Comment (longtimeago) [inline src/test/java/com/jcabi/github/RtForkTest.java]: one more thing, ```SkipBindException``` sounds like an Exception. Better name it as ```SkipBindExceptionRule```

**Comment by longtimeago** [comment]
@cyberone Kind of that :) But:
1) How would you know that you actually fixed the issue?
2) This Rule is going to be used in many places, so we need to be sure it works as expected

By the way, it's name is confusing. See remark above.

**Comment by cyberone** [comment]
@longtimeago 
1a)I reproduced bug in a VM where I captured all available ports (it was terrible), I won't do this at home anymore. 
1b) Try youself this test by giving both MkGrizzlyContainers same port. 
2) How do I trace calls to https://github.com/jcabi/jcabi-log/? Should I extract interface to trace call to the test of test?

**Comment by longtimeago** [comment]
@cyberone let's ask PO if he is OK with current fix?
@yegor256 Are you OK with current fix?

**Comment by yegor256** [comment]
@cyberone @longtimeago technically this looks correct, but it's less elegant that this:

1. Create a new JUnit resource, called `RandomPort`

2. Use in all tests that need a random TCP port

3. Inside that resource catch that BindException and skip the test.

What do you think?

**Comment by longtimeago** [comment]
@yegor256 You approach more explicit ... but it also implies more changes. With current approach we need to add one ```SkipBindExceptionRule``` to test file and that's it.
In your proposal we need to add ```RandomPort``` to test file AND add generated port to ALL tests were ```MkGrizzlyContainer``` is used.
Kind of this:
```
new MkGrizzlyContainer().start(port)
```

**Comment by yegor256** [comment]
@longtimeago in either case we should modify all `*ITCase.java` files, right?

**Comment by longtimeago** [comment]
@yegor256 definitely yes

**Comment by yegor256** [comment]
@longtimeago yes, right, all test files. I think it's better to do it right, if we're modifying all files anyway

**Comment by longtimeago** [comment]
@yegor256 I don't mind

**Comment by cyberone** [comment]
@longtimeago is this way ok or I should implement the pool of ports to give to MkGrizzlyContainers?

**Comment by longtimeago** [comment]
@yegor256 So, what is the decision?
As far as I see, we have 2 options:

1) Create a resource which obtain free port before test
```
public final class ReservePort extends ExternalResource {
    public int port() {
        try (ServerSocket socket = new ServerSocket()) {
            socket.setReuseAddress(true);
            socket.bind(
                new InetSocketAddress(InetAddress.getLoopbackAddress(), 0)
            );
            return socket.getLocalPort();
        } catch (final IOException ex) {
            throw new IllegalStateException(ex);
        }
    }
}
```
and provide that port to ```MkGrizzlyContainers```
```
new MkGrizzlyContainer().start(port)
```
2) Create a resource which wraps the test and skip test in case of ```BindException```like in the PR and use it in all tests

**Comment by yegor256** [comment]
@longtimeago I like option no.1

**Comment by cyberone** [comment]
@yegor256, @longtimeago  is it OK to use 1.7's try(resource allocation) style?

**Comment by longtimeago** [comment]
@yegor256 ok for minimum Java 1.7 for the project?

**Comment by yegor256** [comment]
@longtimeago I don't think so... it's a library, not a tool. We should allow 1.6 users to use it.

**Comment by longtimeago** [comment]
@cyberone my example should be reworked to be in java 1.6 style

**Comment by cyberone** [comment]
@longtimeago here it goes

File: src/test/java/com/jcabi/github/RandomPort.java [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+import org.junit.runners.model.Statement;
+
+/**
+ * Test resource for skipping tests when random port is busy.
+ * @author Aleksey Popov (alopen@yandex.ru)
+ * @version $Id$
+ */
+public class RandomPort extends ExternalResource {
+    @Override
+    public final Statement apply(
```
Comment (longtimeago) [inline src/test/java/com/jcabi/github/RandomPort.java]: we don't need to override this method from ```ExternalResource``` anymore

File: src/test/java/com/jcabi/github/RtForkTest.java [hunk 3]
```diff
@@ -50,6 +51,15 @@
  * @checkstyle ClassDataAbstractionCouplingCheck (500 lines)
  */
 public final class RtForkTest {
+    /**
+     * The rule for skipping test if there's BindException.
+     * @todo 989 Apply this rule to other classes that use MkGrizzlyContainer
```
Comment (longtimeago) [inline src/test/java/com/jcabi/github/RtForkTest.java]: please, use ```#``` before task number
```
@todo #989 Apply this ...
```

**Comment by longtimeago** [comment]
@cyberone please, see 2 remarks

File: src/test/java/com/jcabi/github/RandomPort.java (hunk 2 above)
Comment (cyberone) [inline src/test/java/com/jcabi/github/RandomPort.java]: @longtimeago we need, just in case of BindException

**Comment by cyberone** [comment]
@longtimeago I fixed the puzzle, but are you sure I should fix the handling of Bindxception?

**Comment by longtimeago** [comment]
@cyberone @yegor256 Looks like we've forgotten about the goal of the task - pass the build if no free port available in the system.
Solution 1 (from my proposal) does not solve this, because, anyway, we will get ```IllegalStateException``` instead of ```BindException``` if no free ports available.

Looks like I was wrong, and the only way to achieve the task goal is to intercept and swallow ```BindException``` (as it was proposed initially by PR author).

@yegor256 need your final word

**Comment by yegor256** [comment]
@longtimeago I don't think so. We just need to do `Assume.assumeTrue(false)` instead of throwing `IllegalStateException` and that's it

**Comment by cyberone** [comment]
@yegor256 without running test?

**Comment by yegor256** [comment]
@cyberone exactly

**Comment by cyberone** [comment]
@longtimeago now is everything ok or the exception class is redundant?

**Comment by longtimeago** [comment]
@yegor256 @cyberone If I understand correctly there would be following test case:
1) Some test tries to get the port
2) Port is not available, so ```BindException``` will be thrown inside ```RandomPort.port()``` and it should NOT be caught. 
3) We let it propagates to the test.
4) We will catch it inside ```RandomPort.apply``` and mark the test as passed with ```Assume.assumeTrue(false)```
Have I missed something?

**Comment by yegor256** [comment]
@longtimeago I would rather do `Assume.assumeTrue(false)` inside `RandomPort.port()` if the port can't be allocated

**Comment by longtimeago** [comment]
@yegor256 I'm not really sure that ```AssumptionViolatedException``` (thrown by ```Assume.assumeTrue(false)```) will be managed by junit framework if we do it inside ```RandomPort.port()```

**Comment by cyberone** [comment]
@yegor256 I'm not sure too.
@longtimeago I've removed redundant exception.

**Comment by longtimeago** [comment]
@cyberone as for me,  current implementation is fine

**Comment by cyberone** [comment]
@longtimeago what about merging this fine implementation with the master repository?

**Comment by longtimeago** [comment]
@cyberone there should be final approve from Yegor. Also we need to wait for green light from travis-ci to don't waste rultor resources in case of some problems (for example, checkstyle :)  )

**Comment by cyberone** [comment]
@longtimeago should I merge with master before you ask rultor?

**Comment by longtimeago** [comment]
@cyberone yes, it would be good idea to merge from master to your branch and resolve merge conflicts (if any)

**Comment by cyberone** [comment]
@yegor256 please respond, ok to merge or something should be done?

**Comment by yegor256** [comment]
@rultor try to merge

**Comment by longtimeago** [comment]
@yegor256 we've got communication issue
```
repo.maven.apache.org: Name or service not known
```

**Comment by longtimeago** [comment]
@yegor256 please, retry

**Comment by yegor256** [comment]
@rultor merge

**Comment by longtimeago** [comment]
@yegor256 shouldn't the PR be paid? 8 days passed

**Comment by longtimeago** [comment]
@yegor256 ping

**Comment by longtimeago** [comment]
@yegor256 ping again ... is there something wrong?

**Comment by yegor256** [comment]
@longtimeago will be closed/paid soon

**Comment by longtimeago** [comment]
@yegor256 ok, sorry for annoying

**Comment by longtimeago** [comment]
@yegor256 ping :)

**Comment by dmarkov** [comment]
@longtimeago **32 mins** sent to your balance (ID `52066267`), many thanks!... +32 added to your rating, at the moment it is: [+1976](http://www.netbout.com/b/35024?open=rating)

**Comment by dmarkov** [comment]
@rultor deploy pls

**Pull request merged**